pkg/cloudprovider/providers/openstack/openstack_instances.go

```diff
@@ -81,6 +88,41 @@ func (i *Instances) List(name_filter string) ([]types.NodeName, error) {
 	return ret, nil
 }
 
+// GetAllZones returns all the zones from current servers in same region
+func (os *OpenStack) GetAllZones() (sets.String, error) {
+	// TODO: Caching, but this is currently only called when we are creating a volume,
+	// which is a relatively infrequent operation, and this is only # zones API calls
+	zones := sets.NewString()
+
+	// retrieve all servers
+	opts := servers.ListOpts{}
+	// Retrieve a pager (i.e. a paginated collection)
+	compute, err := openstack.NewComputeV2(os.provider, gophercloud.EndpointOpts{
+		Region: os.region,
+	})
+	if err != nil {
+		glog.Warningf("Failed to find compute endpoint: %v", err)
+		return nil, err
+	}
+	pager := servers.List(compute, opts)
```

Comment:
Hrm.  So this is getting *all* servers in this tenant (and region).  What if the k8s cluster is only using a subset of the servers, and hence this function ends up finding AZs for which there is no k8s node?  Do we care about this?

Unless we can do better, I think we should match the `aws.go` behaviour here (https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/aws/aws.go#L1120).  The aws version seems to do what's in the current code (all zones, regardless of whether they're in the cluster or not), with the additional check that it only looks at running instances, and attempts to skip "master" nodes.

(Better would be to raise this logic out of cloudprovider code, since k8s already internally has the az annotations for all nodes...)